Framework/Algorithms/src/ReflectometryBackgroundSubtraction.cpp

```diff
@@ -0,0 +1,196 @@
+  poly->setProperty("XRanges", getPropertyValue("XRanges"));
+  poly->setProperty("CostFunction", getPropertyValue("CostFunction"));
+  poly->execute();
+  API::MatrixWorkspace_sptr bgd = poly->getProperty("OutputWorkspace");
+
+  // transpose again to get the background
+  transpose->setProperty("InputWorkspace", bgd);
+  transpose->execute();
+  API::MatrixWorkspace_sptr transposedBgd =
+      transpose->getProperty("OutputWorkspace");
+
+  // subtract the background from the input workspace
+  auto output = minus(inputWS, transposedBgd);
+  setProperty("OutputWorkspace", output);
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void ReflectometryBackgroundSubtraction::init() {
+
+  // Input workspace
+  declareProperty(make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input,
+                      boost::make_shared<CommonBinsValidator>()),
+                  "An input workspace.");
+
+  std::vector<std::string> backgroundTypes = {"Per Detector Average",
+                                              "Polynomial"};
+  declareProperty("TypeOfBackgroundSubtraction", "Per Detector Average",
+                  boost::make_shared<StringListValidator>(backgroundTypes),
+                  "The type of background reduction to perform.",
+                  Direction::Input);
+
+  declareProperty(make_unique<ArrayProperty<size_t>>("BottomBackgroundRange",
+                                                     Direction::Input),
+                  "A list of the bottom background ranges.");
+  declareProperty(make_unique<ArrayProperty<size_t>>("TopBackgroundRange",
+                                                     Direction::Input),
+                  "A list of the top background ranges.");
+
+  setPropertyGroup("BottomBackgroundRange",
+                   "Average Per Spectra Background Subtraction");
+  setPropertyGroup("TopBackgroundRange",
+                   "Average Per Spectra Background Subtraction");
+  setPropertySettings(
+      "BottomBackgroundRange",
+      make_unique<EnabledWhenProperty>("TypeOfBackgroundSubtraction",
+                                       IS_EQUAL_TO, "Per Detector Average"));
+  setPropertySettings(
+      "TopBackgroundRange",
+      make_unique<EnabledWhenProperty>("TypeOfBackgroundSubtraction",
+                                       IS_EQUAL_TO, "Per Detector Average"));
+
+  auto increasingAxis = boost::make_shared<IncreasingAxisValidator>();
+  auto nonnegativeInt = boost::make_shared<BoundedValidator<int>>();
+  nonnegativeInt->setLower(0);
+  auto orderedPairs = boost::make_shared<ArrayOrderedPairsValidator<double>>();
+  declareProperty("DegreeOfPolynomial", 0, nonnegativeInt,
+                  "Degree of the fitted polynomial.");
+  declareProperty(make_unique<ArrayProperty<double>>(
```

Comment:
Would it be possible to use `copyProperty` here, similarly to [this](https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/src/CarpenterSampleCorrection.cpp#L57)?